### PR TITLE
[Doc]  #13112 Sprintf timezone behavior

### LINKED ISSUE #13112: Document sprintf timestamp timezone behavior
## Context
Logstash allows for putting dynamic fields in the event data, such as the current time in your specified format.

This format is specified in the [sprintf +FORMAT](https://www.elastic.co/guide/en/logstash/current/event-dependent-configuration.html#sprintf), and applied to the @timestamp field. 

## Issue
The locale of the `sprintf` differs from the locale of the `date` filter. This creates inconsistencies in the rollover of indices and other unexpected behavior. 

#### Difference
**sprintf locale**
https://github.com/elastic/logstash/blob/c698aa224b44c7356fbfeb58ebb0c4a66d72dc71/logstash-core/src/main/java/org/logstash/StringInterpolation.java#L81

**date locale**
The [platform default](https://www.elastic.co/guide/en/logstash/current/plugins-filters-date.html#plugins-filters-date-timezone)

## Moving forward
Suggested to make this clear in the [documentation](https://www.elastic.co/guide/en/logstash/current/event-dependent-configuration.html#sprintf).

### PR DESCRIPTION
**Pull request by SirBouBou**
## Release notes

[m:skip]

## What does this PR do?

This PR try to answer the issue #13112 by explaining a problem which can lead to an unexpected behavior with sprintf.

This commit makes change to docs/static/configuration.asciidoc.

## Why is it important/What is the impact to the user?

This can prevent people encountering an unexpected behavior while using sprintf format.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)
- [ ] I have added tests that prove my fix is effective or that my feature works

## How to test this PR locally

You can view the change in the documentation (Configuring Logstash » Accessing event data and fields in the configuration)
## Related issues

- Closes #13112

### CODE REVIEW

**Comment by logstashmachine** [comment]
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

File: docs/static/configuration.asciidoc [hunk 1]
```diff
@@ -392,6 +392,9 @@ output {
 }
 ----------------------------------
 
+But be careful, the local environment of sprintf differs from the local environment of the date filter. As a result, this can create inconsistencies in the rollover of indices and other unexpected behaviours.
+This is why it is best to specify a canonical time zone ID to be used for date analysis which will avoid any inconsistencies during your use such as http://joda-time.sourceforge.net/apidocs/org/joda/time/format/DateTimeFormat.html[joda time format] or https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/format/DateTimeFormatter.html#patterns[java time format] (recommanded).
+
```
Comment (yaauie) [inline docs/static/configuration.asciidoc:397]: > the local environment of sprintf differs from the local environment of the date filter

Let's talk about the "inconsistencies" you are experiencing.

 - The `@timestamp` holds a `LogStash::Timestamp` object, which by definition an instant on the UTC timeline, so formatting it with sprintf may not yield the same calendar day or clock hour as your machine-local timezone.
 - The Date filter _can_ be configured with a [`timezone`](https://www.elastic.co/guide/en/logstash/current/plugins-filters-date.html#plugins-filters-date-timezone), which it uses as plugin-local context to disambiguate strings that do not contain offset information. Whether or not this is supplied, the resulting timestamp object will be a `LogStash::Timestamp` and therefore an instant on the UTC timeline. The original offset information (as provided by the string or assumed with local context) is not retained.

**Review by yaauie — COMMENTED** [review]
Thank you for taking the time to contribute.

I agree that this is a spot where users can get confused, and the docs could be more clear.

Our timestamp object represents an instant on the continuous, non-overlapping UTC timeline, and uses a UTC-zoned formatter so that we can avoid the surprises of timezones that have overlap-hours (when DST is ending) and skip-hours (when DST is beginning). 

I've taken a stab at making this more clear by explicitly calling out the UTC nature of the Logstash timestamp type in https://github.com/elastic/logstash/pull/13576 -- does this address your concerns from the original #13571 ?

Roping in @karenzone for docs 👀

File: docs/static/configuration.asciidoc (hunk 1 above)
Comment (SirBouBou) [reply docs/static/configuration.asciidoc:397]: Yes, your commit seems to make things more clear, I think it's a great addition to the documentation.

Comment (karenzone) [reply docs/static/configuration.asciidoc:397]: Thanks for sharing your experience and helping us clarify the docs, @SirBouBou.

**Pull request closed without merging**